src/pubsub.c

```diff
@@ -272,15 +271,18 @@ int pubsubSubscribeChannel(client *c, robj *channel, pubsubtype type) {
             slot = getKeySlot(channel->ptr);
         }
 
-        de = kvstoreDictAddRaw(*type.serverPubSubChannels, slot, channel, &existing);
-
-        if (existing) {
-            clients = dictGetVal(existing);
-            channel = dictGetKey(existing);
+        hashtablePosition pos;
+        void *existing;
+        if (!kvstoreHashtableFindPositionForInsert(*type.serverPubSubChannels, slot, channel, &pos, &existing)) {
+            clients = existing;
+            channel = *(robj **)clients->metadata;
         } else {
+            /* Store pointer to channel name in the dict's metadata. */
             clients = dictCreate(&clientDictType);
-            kvstoreDictSetVal(*type.serverPubSubChannels, slot, de, clients);
+            memcpy(clients->metadata, (void *)&channel, sizeof(void *));
```

Comment:
```suggestion
            memcpy(dictMetadata(clients), (void *)&channel, sizeof(void *));
```

Reply:
Also feels like this could benefit from a helper in dict.h.

Reply:
Helper like what? A helper to store a pointer in metadata?

I'd like to replace this dict with a hashtable too, in the future. :) It's an easy change. We'd just use hashtable metadata in the same way. 

Instead of memcpy, we could just store it the same way we read it, like this:

```suggestion
            *(robj **)dictMetadata(clients) = channel;
```